File: 2022/Sources/advent2022/Day15.swift
import Foundation
struct Day15 : Day {

    struct Sensor {
        let sx :Int, sy :Int
        let bx :Int, by :Int

        var range :Int { abs(bx-sx) + abs(by-sy) }
        func addTo (_ sparse :inout SparseRange, _ y :Int) {
            let xrange = range - abs(y-sy)
            if xrange >= 0 { sparse.add((sx-xrange)...(sx+xrange)) }
        }
    }

    struct SparseRange {
        var bounds = 0...0
        var spill = 0...0
        var count = 0

        var covered :Int { span(bounds) + (count > 1 ? span(spill) : 0) }
        var firstGap :Int { count < 2 ? -1 : bounds.upperBound+1 }

        mutating func add (_ range :ClosedRange<Int>) {
            if count == 0 {
                bounds = range
                count = 1
            } else if shouldMerge(bounds, range) {
                bounds = merge(bounds, range)
                if count == 2 && shouldMerge(bounds, spill) {
                    bounds = merge(bounds, spill)
                    count = 1
                }
            } else if count == 1 {
                if range.lowerBound < bounds.lowerBound {
                    spill = bounds
                    bounds = range
                } else {
                    spill = range
                }
                count = 2
            } else if (shouldMerge(spill, range)) {
                spill = merge(spill, range)
                if count == 2 && shouldMerge(bounds, spill) {
                    bounds = merge(bounds, spill)
                    count = 1
                }
            } else {
                print("OVERFLOW!")
            }
        }

        func contains (_ x :Int) -> Bool { bounds.contains(x) || (count > 1 && spill.contains(x)) }

        private func span (_ b :ClosedRange<Int>) -> Int { b.upperBound - b.lowerBound + 1 }
        private func shouldMerge (_ a :ClosedRange<Int>, _ b : ClosedRange<Int>) -> Bool {
            a.overlaps(b) || (a.upperBound+1 == b.lowerBound)
        }
        private func merge (_ a :ClosedRange<Int>, _ b :ClosedRange<Int>) -> ClosedRange<Int> {
            min(a.lowerBound, b.lowerBound) ... max(a.upperBound, b.upperBound)
        }
    }

    let pattern = #"Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)"#
    func parse (_ input :String) throws -> Sensor {
        let regex = try NSRegularExpression(pattern: pattern)
        let nsrange = NSRange(input.startIndex ..< input.endIndex, in: input)
        let matches = regex.matches(in: input, options: [], range: nsrange)
        guard let match = matches.first else { return Sensor(sx: 0, sy: 0, bx: 0, by: 0) }
        func group (_ num :Int) -> Int { Int(input[Range(match.range(at: num), in: input)!])! }
        return Sensor(sx: group(1), sy: group(2), bx: group(3), by: group(4))
    }

    func part1 () throws -> String {
        var sensors = try readInput(15).map(parse)
        sensors.sort(by: { $0.sx < $1.sx })
        let y = 2000000
        var sparse = SparseRange()
        for s in sensors { s.addTo(&sparse, y) }
        let skip = Set(
            sensors.filter({ $0.sy == y && sparse.contains($0.sx) }).map({ $0.sx }) +
            sensors.filter({ $0.by == y && sparse.contains($0.bx) }).map({ $0.bx }))
        return String(sparse.covered - skip.count)
    }

    func part2 () throws -> String {
        var sensors = try readInput(15).map(parse)
        sensors.sort(by: { $0.sx < $1.sx })
        let max = 4000000
        for y in 0 ... max {
            var sparse = SparseRange()
            for s in sensors { s.addTo(&sparse, y) }
            let x = sparse.firstGap
            if x >= 0 && x <= max { return String(x * 4000000 + y) }
        }
        return "?"
    }
}

File: 2022/scala/src/main/scala/Day15.scala
object Day15 extends Day(15):

  case class Bounds (l :Int, u :Int):
    def merge (o :Bounds) = Bounds(Math.min(l, o.l), Math.max(u, o.u))
    def span = u - l + 1
    def contains (v :Int) = l <= v && u >= v
    def overlaps (b :Bounds) = contains(b.l) || b.contains(u)
    def shouldMerge (b :Bounds) = overlaps(b) || (u+1 == b.l)

  case class Sensor (sx :Int, sy :Int, bx :Int, by :Int):
    def range = Math.abs(bx-sx) + Math.abs(by-sy)

  class SparseRange (sensors :Seq[Sensor], y :Int):
    var bounds = Bounds(0, 0)
    var spill = Bounds(0, 0)
    var count = 0

    for (s <- sensors)
      val xrange = s.range - Math.abs(y-s.sy)
      if xrange >= 0 then add(Bounds(s.sx-xrange, s.sx+xrange))

    def covered :Int = bounds.span + (if count > 1 then spill.span else 0)
    def contains (x :Int) = bounds.contains(x) || (count > 1 && spill.contains(x))
    def firstGap :Int = if count < 2 then -1 else bounds.u+1

    def add (b :Bounds) =
      if count == 0 then
        bounds = b
        count = 1
      else if bounds.shouldMerge(b) then
        bounds = bounds.merge(b)
        if count == 2 && bounds.shouldMerge(spill) then
          bounds = bounds.merge(spill)
          count = 1
      else if count == 1 then
        if b.l < bounds.l then
          spill = bounds
          bounds = b
        else spill = b
        count = 2
      else if spill.shouldMerge(b) then
        spill = spill.merge(b)
        if count == 2 && bounds.shouldMerge(spill) then
          bounds = bounds.merge(spill)
          count = 1
      else print("OVERFLOW!")

  val pattern = """Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)""".r
  def parseSensor (input :String) = pattern.findFirstMatchIn(input) match
    case Some(m) => Sensor(m.group(1).toInt, m.group(2).toInt, m.group(3).toInt, m.group(4).toInt)
    case None    => Sensor(0, 0, 0, 0)

  override def answer1 (input :Seq[String]) =
    val sensors = input.map(parseSensor).sortWith(_.sx < _.sx)
    val y = if input.size == 14 then 10 else 2000000
    val sparse = SparseRange(sensors, y)
    val skip = Set(sensors.filter({ s => s.sy == y && sparse.contains(s.sx) }).map(_.sx) ++
                   sensors.filter({ s => s.by == y && sparse.contains(s.bx) }).map(_.bx))
    sparse.covered - skip.size

  override def answer2 (input :Seq[String]) =
    val sensors = input.map(parseSensor).sortWith(_.sx < _.sx)
    val max = if input.size == 14 then 20 else 4000000
    def loop (y :Int) :Long =
      var sparse = SparseRange(sensors, y)
      val x = sparse.firstGap
      if x >= 0 && x <= max then x.toLong * 4000000 + y
      else if y == max then 0
      else loop(y+1)
    loop(0)
